JAMES-3398 add Maven wrapper (with latest Maven version 3.6.3)

aasaru opened the pull request.

chibenwa reviewed (commented):
I think we should not commit binaries here. As a commiter, I can not ensure the authenticity nor the integrity of it. (I would not commit it to the james project as it is)

Your PR don't explain nor document how to use the wrapper (and why). A bit of ducmuentation in developer documentation could be helpful ;-)

Cheers.

aasaru commented:
@chibenwa , fair points both, I agree with you. I now removed the binary and added separate section to readme.

I also improved the readme a bit. I investigated why IntelliJ IDEA complains about Readme.adoc and fixed some of the problems it pointed out (I couldn't fix titles that contain "+" sign so IntelliJ IDEA will still complain about those)

chibenwa commented on .mvn/wrapper/MavenWrapperDownloader.java:
```diff
@@ -0,0 +1,117 @@
+import java.io.*;
+import java.nio.channels.*;
+import java.util.Properties;
+
```
I understand this code is copied but can we have a link to the place it was copied from as a comment?

This helps code patternity.

This helps checking divergences from the original version.

chibenwa commented on mvnw:
```diff
@@ -0,0 +1,310 @@
+# specific language governing permissions and limitations
+# under the License.
+# ----------------------------------------------------------------------------
+
```
Idem, I would welcome a URL from where this was copied.

chibenwa commented on mvnw.cmd:
```diff
@@ -0,0 +1,182 @@
+@REM specific language governing permissions and limitations
+@REM under the License.
+@REM ----------------------------------------------------------------------------
+
```
URL?

chibenwa reviewed (commented):
Sorry for radio silence.

I agree with the proposed changes, thanks for putting that together!

jeantil commented:
A quick search yielded https://github.com/bdemers/maven-wrapper superseded by  https://github.com/takari/maven-wrapper which itself mentions: 

> The project codebase has been accepted to be included in the upstream Apache Maven project itself. Currently the plan is to release the wrapper as a feature of the upcoming Maven 3.7.0 release

jeantil commented:
Note that maven 3.7.x and 3.8.x seem to have been abandonned in favor of maven 4.0.x 
There is an alpha snapshot floating around for fearless osx using people : https://maarten.mulders.it/2020/11/whats-new-in-maven-4/ there is also probably a way to use it on linux...

chibenwa commented:
Is there any plan to carry on this work?

jeantil commented:
Note that 3.8.x was released after all and 4.0.x is still not available. 
in the 3.8.x release, maven wrapper was added as an official maven plugin https://maven.apache.org/wrapper/maven-wrapper-plugin/index.html
